duckduckgo/Android
# Refactor the ViewModelFactory to use multibindings

Task/Issue URL: https://app.asana.com/0/414730916066338/1199916263225062/f
Tech Design URL: 
CC: 

**Description**:
The `ViewModelFactory` is in charge of creating all the view models for the whole app. Before it had to:
- depend on all the dependencies required by the view models
- have visibility of all the view model types
- classes that needed a view model had to have a reference to the view model factory

By using the plugin pattern in this class we:
- decoupled the view model `Factory` class from the actual factories that create the view models
- both factories and the view model `Factory` can now live in different modules if needed without creating circular dependencies

**Steps to test this PR**:
This PR affects pretty much the whole app as it touches all the view models we've got. To test the PR test the common app screens to see if something is amiss.

1. App compiles
2. install/upgrade the app from this branch
3. use the SERP to perform searches
4. form the SERP, open several tabs
5. add/remove bookmarks
6. add/remove fireproof sites
7. navigate to settings screen
8. use fire button
10. clear app data and go through onboarding UX
11. for all above, verify app works as expected


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
